src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+
+        public bool Remove(TKey key, out TValue value)
+        {
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                value = _entries[index].Value;
+                RemoveAt(index);
+                return true;
+            }
+            value = default;
+            return false;
+        }
+
+        public void RemoveAt(int index)
+        {
+            int count = Count;
+            if (index < 0 || index >= count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
+            }
+
+            // Remove the entry from the bucket
+            UpdateBucket(index, true);
+
+            // Decrement the indices > index
+            for (int i = index + 1; i < count; ++i)
+            {
+                UpdateBucket(i, false, -1);
+            }
+            Entry[] entries = _entries;
+            Array.Copy(entries, index + 1, entries, index, count - index - 1);
+            --Count;
+            entries[Count] = default;
+            ++_version;
+        }
+
+        public void SetAt(int index, TValue value)
+        {
+            if (index < 0 || index >= Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index), "Index was out of range. Must be non-negative and less than the size of the collection.");
+            }
+
+            _entries[index].Value = value;
+            ++_version;
+        }
+
+        public void SetValue(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.OverwriteExisting);
+
+        public void TrimExcess() => TrimExcess(Count);
+
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            int newSize = HashHelpers.GetPrime(capacity);
+            if (newSize < _entries.Length)
+            {
+                Resize(newSize);
+                ++_version;
+            }
+        }
+
+        public bool TryAdd(TKey key, TValue value) => TryInsert(null, key, value, InsertionBehavior.None);
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            int index = IndexOf(key);
+            if (index >= 0)
+            {
+                value = _entries[index].Value;
+                return true;
+            }
+            value = default;
+            return false;
+        }
+
+        #region Explicit Interface Implementation
+        KeyValuePair<TKey, TValue> IList<KeyValuePair<TKey, TValue>>.this[int index]
+        {
+            get
+            {
+                TValue value = GetAt(index, out TKey key);
+                return new KeyValuePair<TKey, TValue>(key, value);
+            }
+            set => throw new NotSupportedException(); // Should this be supported?
```

Comment:
Thoughts on this?

Reply:
I think it's fine to throw for now, but it should be InvalidOperationException.

Reply:
Scratch that, NotSupportedException is probably better.

Reply:
I ended up implementing this.